src/CanvasRenderingContext2d.cc

```diff
@@ -330,6 +402,9 @@ Context2d::fill(bool preserve) {
       ? shadow(cairo_fill)
       : cairo_fill(_context);
   }
+  if (new_pattern != state->fillPattern && new_pattern != NULL) {
```

Comment:
I think the `new_pattern != state->fillPattern` part of this makes the `cairo_pattern_destroy` unreachable in cases of global alpha < 1. Here again the `destroy` fn actually just decreases ref count; it should be safe to call it whenever `new_pattern != NULL`.

Actually, you can make the `destroy` call after line 388 I think, right after `cairo_set_source(_context, new_pattern)`. That'll let you shrink the scope of `new_pattern` as well.

Reply:
i added this check because the new_pattern can be NULL globalAlpha = 1 or can be fillPattern if the creation of the new pattern failed ( because we return the same pattern from the other function ).
Why you think is unreachable?